flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -682,7 +682,9 @@ private void stopTaskExecutorServices() throws Exception {
             try {
                 changelogStorage =
                         changelogStoragesManager.stateChangelogStorageForJob(
-                                jobId, taskManagerConfiguration.getConfiguration());
+                                jobId,
+                                taskManagerConfiguration.getConfiguration(),
+                                taskManagerMetricGroup.getJobMetricsGroup(jobId));
```

Comment:
Could you elaborate why this should be changed?

To me, it doesn't seem directly related to my change; I also don't see any obvious reasons why one way is better than the other.

Reply:
On the one hand it is about consistency; we only use getters for things like the IOMetricGroups, because there we can guarantee that they actually exists because they are created in the constructor of the containing group.

Here, we rely on some other call above that is _technically_ not quite related to the changelog storage metrics to implicitly create the TMJMG. To someone who doesn't know how `addTaskForJob` works (aka, that it even creates 2 groups), this isn't intuitive.

Doing 2 steps at once was fine previously because no one needed the TMJMG anyway, but this has changed now. It should thus be split.

Reply:
Thanks for clarifying. I''ve added two commits to split it (for production and test code, to be squashed into one upon merging).